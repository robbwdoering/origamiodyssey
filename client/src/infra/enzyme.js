/**
 * FILENAME: enzyme.js
 *
 * DESCRIPTION: Setup for AirBnB's Enzyme package. 
 */

import Enzyme, { configure, shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
export { shallow, mount, render };
export default Enzyme;
