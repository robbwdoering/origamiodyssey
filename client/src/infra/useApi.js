/**
 * FILENAME: useApi.js
 *
 * DESCRIPTION: Helper hook that manages a connection to the authenticated REST API that backs the app.
 * src: https://github.com/auth0/auth0-react/blob/master/EXAMPLES.md#4-create-a-useapi-hook-for-accessing-protected-apis-with-an-access-token
 * @param path JUST THE PATH, not the full URL, of the requested resource
 * @param options object defining audience, scope, and any custom inputs
 */
import { useState } from 'react';
import { useAuth0 } from '@auth0/auth0-react';

export const useApi = (url, method, options = {}, handleFetch, data) => {
    const { getAccessTokenSilently, user, isAuthenticated } = useAuth0();
    const [state, setState] = useState({
        error: null,
        loading: false,
        data: null
    });

    const sleep = ms => {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    const perform = async (urlSuffix, overrideValues, overrideMethod) => {
        let count = 20;
        while ((!isAuthenticated) && count > 0) {
            count--;
            await sleep(1000);
        }

        try {
            const accessToken = await getAccessTokenSilently(options);

            let msg = {
                method: overrideMethod || method,
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                    // Authorization: `Bearer ${accessToken}`,
                    strategy: 'auth0',
                    access_token: accessToken
                }
            };

            // Append data if it was passed
            if (overrideValues) {
                msg.body = JSON.stringify(overrideValues);
            } else if (data && method !== "GET") {
                msg.body = JSON.stringify(data);
            }
            setState(Object.assign({}, state, { loading: true }))

            // Perform the fetch asynchrounously
            fetch(process.env.REACT_APP_API_IP + url + (urlSuffix || ""), msg)
            .then(response => {
                if (!response.ok) {
                    console.error('[useApi] Received a bad status from the server: ', response);
                    throw new Error(`status ${response.status}`);
                }
                return response.json();
            })
            .then((json) => {
                // This line handles all responses - created at app root to hook into redux
                if (handleFetch) {
                    handleFetch(json);
                }

                setState(Object.assign({}, state, {data: json}))

                // Turn off loading indicator
                setState(Object.assign({}, state, { loading: false }));
            })
            .catch(e => {
                console.log('call failed', e);
            });
        } catch (error) {
            console.log('useApi FAILURE from exception: ', error);
            setState({
                ...state,
                error,
                loading: false
            });
        }
    };

    return {
    ...state,
    refresh: perform
    };
};
