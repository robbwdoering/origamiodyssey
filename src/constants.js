/**
 * FILENAME: Constants.js 
 *
 * DESCRIPTION: Contains constants for use accross the app. 
 */

export const Pages = {
	Splash: "Splash",
	Lesson: "Lesson"
};

export const initAppReducerState = {
	page: Pages.Splash
};
