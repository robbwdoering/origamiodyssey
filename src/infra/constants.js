/**
 * FILENAME: Constants.js 
 *
 * DESCRIPTION: Contains constants for use accross the app. 
 */

import React, { useState, useRef, useMemo, useEffect } from 'react';
import GitHubIcon from "@material-ui/icons/GitHub";
import SchoolIcon from "@material-ui/icons/School";
import ClassIcon from "@material-ui/icons/Class";
import DashboardIcon from "@material-ui/icons/Dashboard";
import GroupAddIcon from "@material-ui/icons/GroupAdd";

// The FOLD files 
import BirdBase from "./../folds/BirdBase.json";
import BoatBase from "./../folds/BoatBase.json";
import FrogBase from "./../folds/FrogBase.json";
import SailBoat from "./../folds/SailBoat.json";

export const Pages = {
	Splash: "Splash",
	ModelSelect: "ModelSelect",
	Fold: "Fold",
	User: "User",

	// Used but undefined
	Lesson: "Lesson",
	History: "History",
	LearnDashboard: "LearnDashboard",
	TeachDashboard: "TeachDashboard"
};

export const Tags = {
	base: {
		text: "Base",
		category: "family"
	},
	bird: {
		text: "Bird",
		category: "counterpart",
	},
	boat: {
		text: "Boat",
		category: "counterpart"
	},
	flower: {
		text: "Flower",
		category: "counterpart"
	}
};

export const Folds = {
	BirdBase: {
		name: "Bird Base",
		description: "The first base many learn, this is perfect for winged creations such as birds and dragons.",
		tags: ["base"],
		author: "Traditional",
		json: BirdBase
	},
	BoatBase: {
		name: "Boat Base",
		description: "A more specific base for blocky models, such as boats and frames.",
		tags: ["base"],
		author: "Traditional",
		json: BoatBase
	},
	FrogBase: {
		name: "Frog Base",
		description: "A complex base used for models with length-wise asymmetry, like flowers and animals.",
		tags: ["base"],
		author: "Traditional",
		json: FrogBase 
	},
	SailBoat: {
		name: "Sail Boat",
		description: "An assymetrical sailboat with a built in kickstand - perfect for display.",
		tags: ["boat"],
		author: "Traditional",
		json: SailBoat 
	}
};

export const initAppReducerState = {
	layoutState: {
		hash: 0,
		page: Pages.ModelSelect,
		curFold: null,
		foldHash: 0,
		useImages: false,
		showEditor: true,
		expandHierarchy: false 
	},
	foldState: {
		hash: 0,
		selectedLevel: 0,
		stepIdx: -1, // Depends on selected level
		overrideWithEditor: false,
		usingDefaults: true
	},
	editorState: {
		hash: 0,
		edgeHighlights: [],
		vertexHighlights: [],
		faceHighlights: [],
		showEdges: true,
		showVertices: true,
		showFaces: true,
		showTriangulations: false,
		showLables: false
	},
	showNavDrawer: false

};

export const Actions = {
	SET_LAYOUT_STATE: "SET_LAYOUT_STATE",
	SET_SHOW_NAV_DRAWER: "SET_SHOW_NAV_DRAWER",
	SET_PAGE_CONFIG: "SET_PAGE_CONFIG",
	SET_FOLD_STATE: "SET_FOLD_STATE",
	SET_EDITOR_STATE: "SET_EDITOR_STATE",
};

export const initNavTree = [
	{	
		text: "What is Origami?",
		key: "context_root",
		children: [
			{
				text: "History",
				key: "history"				
			},
			{
				text: "Origami Today",
				key: "current_origami"
			},
			{
				text: "Practical Origami",
				key: "practical"
			}
		]
	},
	{	
		text: "Learn Origami",
		key: "learn_root",
		children: [
			{
				icon: <DashboardIcon />,
				text: "Dashboard",
				key: "learn_dashboard"				
			},
			{
				icon: <GroupAddIcon />,
				text: "Join Class",
				key: "join_class"
			}
		]
	},
	{	
		text: "Teach Origami",
		key: "teach_root",
		children: [
			{
				icon: <DashboardIcon />,
				text: "Lesson Plans",
				key: "lesson_plan_dashboard"				
			},
			{
				icon: <SchoolIcon />,
				text: "Read Research Paper",
				key: "google_scholar"
			},
			{
				icon: <ClassIcon />,
				text: "Origami in the Classroom",
				key: "teaching_tips"
			}
		]
	},
	{	
		icon: <GitHubIcon />,
		text: "Source Code",
		key: "github"
	},
];

export const initFoldState = {
	//place to store buffer geo vertex data
	positions: [],
	//place to store buffer geo vertex colors
	colors: [],
	indices: [],
	nodes: [], 
	faces: [], 
	edges: [], 
	creases: [], 
	vertices: [], //indexed vertices array
	fold: null,
	creaseParams: null
}

/*
export const Lines = {
    U: hingeLines,
    M: mountainLines,
    V: valleyLines,
    C: cutLines,
    F: facetLines,
    B: borderLines
};
*/
