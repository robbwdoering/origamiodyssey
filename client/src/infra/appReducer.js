/**
 * FILENAME: appReducer.js
 *
 * DESCRIPTION: Handles state update for all layout actions.
 */

import supermemo2 from 'supermemo2';
import { initAppReducerState, Actions } from './constants';

const finalInitState = JSON.parse(JSON.stringify(initAppReducerState));

export const appReducer = (state = finalInitState, action) => {
	let newState = Object.assign({}, state);

	switch (action.type) {
		case Actions.SET_SHOW_NAV_DRAWER:
			// If passed a val set to that, otherwise toggle
			newState.showNavDrawer = action.payload !== undefined ? action.payload : !newState.showNavDrawer;
			break;
		case Actions.SET_LAYOUT_STATE:
			console.log('[SET_LAYOUT_STATE]', action.payload);
			Object.assign(newState.layoutState, action.payload || initAppReducerState.layoutState);
			newState.layoutState.hash++;
			break;
		case Actions.SET_FOLD_STATE:
			console.log('[SET_FOLD_STATE]', action.payload);
			Object.assign(newState.foldState, action.payload || initAppReducerState.foldState);
			newState.foldState.hash++;
			break;
		case Actions.SET_EDITOR_STATE:
			console.log('[SET_EDITOR_STATE]', action.payload);
			Object.assign(newState.editorState, action.payload || initAppReducerState.editorState);
			newState.editorState.hash++;
			break;
		case Actions.SET_USER_STATE:
			console.log('[SET_USER_STATE]', action.payload);
			Object.assign(newState.userState, action.payload || initAppReducerState.userState);
			newState.userState.hash++;
			break;
		case Actions.ADD_HISTORY_ENTRY:
			if (!action.payload) {
				return newState;
			}
			newState.userState.foldHistory.push(action.payload)

			// If this model type is part of the memo algorithm, update that entry 
			const idx = newState.userState.modelList.findIndex(model => model.foldKey === action.payload.foldKey);
			if (idx !== -1) {
				console.log("Modifying entry for ", action.payload.foldKey);

				// Update the spaced learning params given this new information
				const quality = action.payload.quality;
				const lastSchedule = newState.userState.modelList[idx].schedule
				const lastFactor = newState.userState.modelList[idx].factor
				newState.userState.modelList[idx] = {
					foldKey: action.payload.foldKey,
					...supermemo2(quality, lastSchedule, lastFactor)
				}
			}
			// Else don't rememeber anything other than the foldHistory row

			newState.userState.hash++;
			console.log('[ADD_HISTORY_ENTRY]', action.payload, newState.userState);
			break;
	}

	return newState;
};
