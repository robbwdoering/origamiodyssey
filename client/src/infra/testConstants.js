/**
 * FILENAME: testConstants.js
 *
 * DESCRIPTION: Contains constants for use in the test suites. 
 */

import { initAppReducerState } from "./constants";

export const testRedux = Object.assign({}, initAppReducerState, {
    foldState: Object.assign({}, initAppReducerState.foldState, {
        maxSteps: 10
    })
});
