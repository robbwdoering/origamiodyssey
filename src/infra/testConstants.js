/**
 * FILENAME: testConstants.js
 *
 * DESCRIPTION: Contains constants for use in the test suites. 
 */

import { initAppReducerState } from "./constants";

export const testRedux = Object.assign({}, initAppReducerState, {});
