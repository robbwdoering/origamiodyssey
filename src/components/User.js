/**
 * FILENAME: User.js
 *
 * DESCRIPTION: This page allows the user to browser through cards, read details on models, and select one to fold.
 */

// React + Redux
import React, { useState, useRef, useMemo, useEffect, createRef } from 'react';
import { connect } from 'react-redux';

import { useUpdate, useSpring, useSprings, animated, config } from 'react-spring';

import {
	Menu,
	ButtonGroup,
	Button,
	Input,
	MenuItem,
	List,
	Divider,
	Table,
	TableBody,
	TableCell,
	TableContainer,
	TableHead,
	TableRow,
	ListItem,
	Grid,
	Typography,
	Paper,
	Tabs,
	Tab
} from '@material-ui/core';
import { ToggleButton } from '@material-ui/lab';
import Remove from '@material-ui/icons/Remove';
import ExitToApp from '@material-ui/icons/ExitToApp';
import SquareLoader from 'react-spinners/SquareLoader';

import { useAuth0 } from '@auth0/auth0-react';

import useStyles from './../style/theme';
import { Folds, Pages } from './../infra/constants';
import { setUserState, setLayoutState } from './../infra/actions';
import { timerPosixToString } from './../infra/utils';

export const User = props => {
	const { userState, setUserState, layoutState, setLayoutState } = props;
	const classes = useStyles();
	const { user, loginWithRedirect, isLoading, logout, isAuthenticated } = useAuth0();
	const [addModelAnchor, setAddModelAnchor] = useState();
	const [assistantTab, setAssistantTab] = useState(0);

	// ----------
	// STATE INIT
	// ----------

	// ----------------
	// MEMBER FUNCTIONS
	// ----------------

	const ControlRow = ({ text, children, ...rest }) => (
		<Grid item className={classes.editor_row} {...rest}>
			{/* Title */}

			<Typography className={classes.modelCard_label} variant='body2' color='textSecondary' component='h4'>
				{text}
			</Typography>
			<Divider />

			{children}
		</Grid>
	);

	const handleUserFormChange = (name, value) => {
		setUserState({ [name]: value });
	};

	const handleModelSelection = foldKey => {
		setAddModelAnchor(null);

		// Exit early if this is already in the list
		if (userState.modelList.find(entry => entry.foldKey === foldKey)) {
			return;
		}

		let newModelListEntry = {
			foldKey: foldKey,
			schedule: 0,
			factor: 2.5,
			isRepeatAgain: true
		};

		const newModelList = [...userState.modelList, newModelListEntry];
		setUserState({ modelList: newModelList });
	};

	const handleOpenFold = foldKey => {
		setLayoutState({
			page: Pages.Fold,
			curFold: foldKey
		});
	};

	const handleRemoveFold = foldKey => {
		const idx = userState.modelList.findIndex(entry => entry.foldKey === foldKey);
		console.log('[handleRemoveFold]', idx);
		// Exit early if this isn't already in the list
		if (idx === -1) {
			return;
		}

		let newArr = [...userState.modelList];
		newArr.splice(idx, 1);

		setUserState({ modelList: newArr });
	};

	const renderAssistantTab = assistantTab => {
		switch (assistantTab) {
			case 0:
				return (
					<React.Fragment>
						<Button className={classes.user_add_model_button} onClick={event => setAddModelAnchor(event.currentTarget)}>
							Add Model To Plan
						</Button>
						<Menu
							name='modelAdd'
							input={<Input id='Menu-multiple-chip' />}
							className={classes.user_add_model_menu}
							anchorEl={addModelAnchor}
							open={Boolean(addModelAnchor)}
							onClose={() => setAddModelAnchor(null)}
						>
							{Object.keys(Folds)
								.filter(foldKey => !userState.modelList.find(entry => entry.foldKey === foldKey))
								.map(foldKey => (
									<MenuItem key={foldKey} value={foldKey} onClick={() => handleModelSelection(foldKey)}>
										{Folds[foldKey].name}
									</MenuItem>
								))}
						</Menu>

						<TableContainer>
							<Table size='small'>
								<TableHead className={classes.user_models_header}>
									<TableRow>
										<TableCell className={classes.slimCol}> Controls </TableCell>
										<TableCell> Model Name </TableCell>
										<TableCell align='right' className={classes.slimCol}>
											{' '}
											Day to Practice{' '}
										</TableCell>
										<TableCell align='right' className={classes.slimCol}>
											{' '}
											Factor{' '}
										</TableCell>
										<TableCell align='right' className={classes.slimCol}>
											{' '}
											In Schedule?{' '}
										</TableCell>
									</TableRow>
								</TableHead>
								<TableBody>
									{userState.modelList.map(entry => {
										console.log('[modelList]', entry);
										const foldObj = Folds[entry.foldKey];
										return (
											<TableRow>
												<TableCell className={classes.slimCol}>
													<ButtonGroup className={classes.user_row_controls} color='primary'>
														<Button onClick={() => handleRemoveFold(entry.foldKey)} title='Remove this model from the schedule'>
															{' '}
															<Remove />{' '}
														</Button>
														<Button onClick={() => handleOpenFold(entry.foldKey)} title='Start this fold'>
															{' '}
															<ExitToApp />{' '}
														</Button>
													</ButtonGroup>
												</TableCell>
												<TableCell> {foldObj.name} </TableCell>
												<TableCell align='right' className={classes.slimCol}>
													{' '}
													{new Intl.DateTimeFormat().format(new Date(Date.now() + entry.schedule * 24 * 60 * 60 * 1000))}{' '}
												</TableCell>
												<TableCell align='right' className={classes.slimCol}>
													{' '}
													{entry.factor.toFixed(2)}{' '}
												</TableCell>
												<TableCell align='right' className={classes.slimCol}>
													{' '}
													{entry.isRepeatAgain ? 'True' : 'False'}{' '}
												</TableCell>
											</TableRow>
										);
									})}
								</TableBody>
							</Table>
						</TableContainer>
					</React.Fragment>
				);
			case 1:
				return (
					<TableContainer>
						<Table size='small'>
							<TableHead className={classes.user_models_header}>
								<TableRow>
									<TableCell align='right' className={classes.slimCol}>
										{' '}
										Controls{' '}
									</TableCell>
									<TableCell> Model Name </TableCell>
									<TableCell align='right' className={classes.slimCol}>
										{' '}
										Day{' '}
									</TableCell>
									<TableCell align='right' className={classes.slimCol}>
										{' '}
										Quality{' '}
									</TableCell>
									<TableCell align='right' className={classes.slimCol}>
										{' '}
										Timer{' '}
									</TableCell>
								</TableRow>
							</TableHead>
							<TableBody>
								{userState.foldHistory.map(entry => {
									const foldObj = Folds[entry.foldKey];
									return (
										<TableRow>
											<TableCell align='right' className={classes.slimCol}>
												<ButtonGroup className={classes.user_row_controls} color='primary'>
													<Button onClick={() => handleOpenFold(entry.foldKey)} title='Start this fold'>
														{' '}
														<ExitToApp />{' '}
													</Button>
												</ButtonGroup>
											</TableCell>
											<TableCell component='th' scope='row'>
												{' '}
												{foldObj.name}{' '}
											</TableCell>
											<TableCell align='right' className={classes.slimCol}>
												{' '}
												{new Intl.DateTimeFormat().format(new Date(entry.time))}{' '}
											</TableCell>
											<TableCell align='right' className={classes.slimCol}>
												{' '}
												{entry.quality + 1}/5{' '}
											</TableCell>
											<TableCell align='right' className={classes.slimCol}>
												{' '}
												{entry.timer ? timerPosixToString(entry.timer) : ''}{' '}
											</TableCell>
										</TableRow>
									);
								})}
							</TableBody>
						</Table>
					</TableContainer>
				);

			case 2:
			default:
				return <span> ERROR </span>;
		}
	};

	// ---------
	// LIFECYCLE
	// ---------

	if (isLoading) {
		return (
			<Typography className={classes.modelCard_label} variant='h3' color='textSecondary' component='h3'>
				<SquareLoader color='#e0e0e0' size={100} />
			</Typography>
		);
	} else if (!isAuthenticated) {
		return (
			<Paper className={classes.user_login_prompt} elevation={3}>
				<Typography className={classes.modelCard_label} variant='h3' color='textSecondary' component='h3'>
					Login
				</Typography>

				<Typography variant='body' color='textSecondary' component='body'>
					Please login to access user page features and preferences.
				</Typography>

				<Button onClick={() => loginWithRedirect()} color='primary'>
					Login
				</Button>
			</Paper>
		);
	}

	console.log('[User]', user);

	return (
		<Grid className={classes.user_container} container spacing={2}>
			{/* Profile Information */}
			<Grid item xs={12} md={4}>
				<Paper className={classes.user_profile} elevation={3}>
					<Typography className={classes.modelCard_label} variant='h3' color='textSecondary' component='h3'>
						Profile
					</Typography>
					<Divider />
					<br />
					<Grid container>
						<ControlRow text='Email' xs={12} md={8}>
							<div> {user.email} </div>
						</ControlRow>
						<ControlRow text='Logout' xs={12} md={4}>
							<Button onClick={() => loginWithRedirect()} color='primary'>
								Logout
							</Button>
						</ControlRow>
					</Grid>
				</Paper>
			</Grid>

			{/* Preferences */}
			<Grid item xs={12} md={8}>
				<Paper className={classes.user_pref} elevation={2}>
					<Typography className={classes.modelCard_label} variant='h3' color='textSecondary' component='h3'>
						Preferences
					</Typography>
					<Divider />
					<br />
					<Grid container spacing={2}>
						<Grid item xs={12} md={4}>
							<ToggleButton
								value={userState.showEditor}
								selected={userState.showEditor}
								onChange={() => handleUserFormChange('showEditor', !userState.showEditor)}
							>
								Show Editor
							</ToggleButton>
						</Grid>
						<Grid item xs={12} md={8}>
							<Typography variant='body2' color='textSecondary' component='h4'>
								Show special cards during instructions for controlling the animation, and identifying individual vertices and edges in the
								paper.
							</Typography>
						</Grid>

						<Grid item xs={12} md={4}>
							<ToggleButton
								value={userState.showTimerAssess}
								selected={userState.showTimerAssess}
								onChange={() => handleUserFormChange('showTimerAssess', !userState.showTimerAssess)}
							>
								Show Timer
							</ToggleButton>
						</Grid>
						<Grid item xs={12} md={8}>
							<Typography variant='body2' color='textSecondary' component='h4'>
								Show a timer that allows you to track model folding speed over time. Turning this on allows the assistant to identify your
								strengths and weakenesses better.
							</Typography>
						</Grid>

						<Grid item xs={12} md={4}>
							<ToggleButton
								value={userState.showLikertAssess}
								selected={userState.showLikertAssess}
								onChange={() => handleUserFormChange('showLikertAssess', !userState.showLikertAssess)}
							>
								Show Quality Prompts
							</ToggleButton>
						</Grid>
						<Grid item xs={12} md={8}>
							<Typography variant='body2' color='textSecondary' component='h4'>
								Show a prompt after completing a model, allowing you to track quality over time. Turning this on allows the assistant to
								identify your strengths and weakenesses better.
							</Typography>
						</Grid>
					</Grid>
				</Paper>
			</Grid>

			{/* Assistant */}
			<Grid item xs={12}>
				<Paper className={classes.user_assistant} elevation={2}>
					{/*					<Typography className={classes.modelCard_label} variant="h3" color="textSecondary" component="h3">
						Assistant
					</Typography>
*/}
					<Tabs value={assistantTab} indicatorColor='primary' textColor='primary' onChange={(e, val) => setAssistantTab(val)}>
						<Tab label='Plan' />
						<Tab label='History' />
						<Tab label='Settings' />
					</Tabs>

					<Divider />
					<br />

					{renderAssistantTab(assistantTab)}
				</Paper>
			</Grid>
		</Grid>
	);
};

export const mapStateToProps = (state, props) => {
	return {
		userState: state.appReducer.userState,
		userStateHash: state.appReducer.userState.hash,
		layoutState: state.appReducer.layoutState,
		layoutStateHash: state.appReducer.layoutState.hash
	};
};

export default connect(mapStateToProps, { setUserState, setLayoutState })(User);
