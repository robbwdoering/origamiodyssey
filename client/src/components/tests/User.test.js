/**
 * FILENAME: User.test.js
 */

// React + Enzyme
import React from 'react';
import { shallow, loginWithRedirect, logout } from './../../infra/enzyme';

// Local
import { testRedux } from './../../infra/testConstants';
import { Folds } from './../../infra/constants';
import { User } from './../User';

describe('User Page', () => {
	let comp;
	let setUserState, setLayoutState;

	beforeEach(() => {
		setUserState = jest.fn();
		setLayoutState = jest.fn();

		comp = shallow(
			<User
				// Redux
				userState={Object.assign({}, testRedux.userState)}
				userStateHash={0}
				layoutState={Object.assign({}, testRedux.layoutState)}
				layoutStateHash={0}
				setUserState={setUserState}
				setLayoutState={setLayoutState}
			/>
		);
	});

	it('renders without crashing', () => {
		expect(comp).toMatchSnapshot();
	});

	it('renders menu options', () => {
		// Open the menu
		comp.find('#oo-assistant-add-model').simulate('click', { currentTarget: {} });
		const menu = comp.find({ name: 'modelAdd' }).first();
		expect(menu.exists()).toBe(true);
		expect(menu.prop('open')).toBe(true);

		expect(menu).toMatchSnapshot();
	});

	// this is a long one, necessary to smoothly test adding and deleting
	it('modifies the the assistant list', () => {
		// Open the menu
		comp.find('#oo-assistant-add-model').simulate('click', { currentTarget: {} });
		const menu = comp.find({ name: 'modelAdd' }).first();

		// Add a model from it
		menu.find('WithStyles(ForwardRef(MenuItem))').first().simulate('click');

		expect(setUserState.mock.calls.length).toBe(1);
		const newEntry = setUserState.mock.calls[0][0].modelList[0];
		expect(newEntry.foldKey).toBe('SailBoat');

		// Use the object to actually change the props
		comp.setProps({
			userState: Object.assign({}, testRedux.userState, { modelList: [newEntry] })
		});

		// Remove the added model
		comp.find({ title: 'Remove this model from the schedule' }).first().simulate('click');

		expect(setUserState.mock.calls.length).toBe(2);
		expect(setUserState.mock.calls[1][0].modelList.length).toBe(0);
	});

	it('renders history tab without crashing', () => {
		comp.find('WithStyles(ForwardRef(Tabs))').prop('onChange')({}, 1);
		comp.update();
		expect(comp.find('#oo-assistant')).toMatchSnapshot();
	});

	it('renders settings tab without crashing', () => {
		comp.find('WithStyles(ForwardRef(Tabs))').prop('onChange')({}, 2);
		comp.update();
		expect(comp.find('#oo-assistant')).toMatchSnapshot();
	});

	it('renders login page', () => {
		comp.setProps({
			logoutOverride: true
		});
		comp.update();

		expect(comp.exists('#oo-assistant')).toBe(false);
		expect(comp).toMatchSnapshot();
	});

	it('changes preferences when toggles are clicked', () => {
		// Click every toggle button
		const toggles = comp.find('#oo-preferences-container').find({ value: false, selected: false });
		for (let i = 0; i < toggles.length; i++) {
			toggles.at(i).prop('onChange')();
		}

		expect(toggles.length).toBe(3);
		expect(setUserState.mock.calls.length).toBe(toggles.length);
		expect(Object.keys(setUserState.mock.calls[toggles.length - 1][0])[0]).toBe('showLikertAssess');
	});
});
