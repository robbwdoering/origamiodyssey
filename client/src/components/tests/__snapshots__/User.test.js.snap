// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Page renders without crashing 1`] = `ShallowWrapper {}`;
